src/Accelerators/NNPA/Transform/ZHigh/ZHighClipToDLFloat.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------- ZHighClipToDLFloat.cpp - ZHigh High Level Optimizer -------===//
+//
+// Copyright 2023- The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file implements a set of rewritten rules to clip CPU numerical values
+// before passing to ZHighStick, which avoids data range violation error due to
+// the dlfloat range.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "src/Accelerators/NNPA/Dialect/ZHigh/ZHighOps.hpp"
+#include "src/Accelerators/NNPA/Dialect/ZHigh/ZHighOps/OpHelper.hpp"
+#include "src/Accelerators/NNPA/Pass/NNPAPasses.hpp"
+#include "src/Dialect/ONNX/DialectBuilder.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Support/TypeUtilities.hpp"
+
+using namespace mlir;
+using namespace onnx_mlir;
+using namespace onnx_mlir::zhigh;
+
+namespace onnx_mlir {
+namespace zhigh {
+
+namespace {
+
+/// Check if a value is from or transitively from a zTensor without value
+/// modification.
+bool valueFromZTensor(Value tensor) {
+  if (!tensor.dyn_cast<BlockArgument>()) {
+    Operation *op = tensor.getDefiningOp();
+    // From a zTensor.
+    if (isa<ZHighUnstickOp>(op))
+      return true;
+    // Concat/Transpose does not change the input precision. So we can consider
+    // that the input is already in the dlfloat range if it comes from zTensor.
+    if (isa<ONNXConcatOp, ONNXTransposeOp>(op)) {
+      return llvm::all_of(
+          op->getOperands(), [&](Value v) { return valueFromZTensor(v); });
+    }
+  }
+  return false;
+}
+
+class ZHighClipToDLFloatPattern : public OpRewritePattern<ZHighStickOp> {
+public:
+  using OpRewritePattern<ZHighStickOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ZHighStickOp stickOp, PatternRewriter &rewriter) const override {
+    Operation *genericOp = stickOp.getOperation();
+    Location loc = genericOp->getLoc();
+
+    Value input = stickOp.getIn();
+    Type inputElementType = getElementType(input.getType());
```

Comment:
Is there a need to explicitly test if the value is a float, and if not, not add the max?

Reply:
The input of stick is float itself. However, we can have a check or an assert here. I'll update this.

Reply:
Added a check for float, only clip if bitwidth is > 16. Thanks!